apng.cpp

```diff
@@ -0,0 +1,457 @@
+#include "apng.hpp"
+
+#include <stdint.h>
+#include <stdio.h>
+#include <png.h>
+
+struct apng_frame_header {
+    uint32_t width;
+    uint32_t height;
+    uint32_t x_offset;
+    uint32_t y_offset;
+    uint16_t delay_num;
+    uint16_t delay_den;
+    uint8_t dispose_op;
+    uint8_t blend_op;
+};
+
+struct apng_decoder_struct {
+    png_structp png_ptr;
+    png_infop info_ptr;
+    const cv::Mat* mat;
+    size_t read_pos;
+    uint32_t frame_num;
+    struct apng_frame_header frame_header;
+    uint8_t *prev_frame;
+    png_bytepp prev_rows;
+};
+
+struct apng_encoder_struct {
+    png_structp png_ptr;
+    png_infop info_ptr;
+    uint8_t *buf;
+    size_t buf_size;
+    size_t write_pos;
+    cv::Mat* prev_frame;
+};
+
+void user_read_data(png_structp png_ptr,
+        png_bytep data, png_size_t length) {
+    apng_decoder d = (apng_decoder) png_get_io_ptr(png_ptr);
+    if (d->read_pos + length > d->mat->total()) {
+        png_error(png_ptr, "Tried to read PNG data past end of file");
+    }
+    memcpy(data, &d->mat->data[d->read_pos], length);
+    d->read_pos += length;
+}
+
+void user_write_data(png_structp png_ptr,
+        png_bytep data, png_size_t length) {
+    apng_encoder e = (apng_encoder) png_get_io_ptr(png_ptr);
+    if (e->write_pos + length >= e->buf_size) {
+        png_error(png_ptr, "Tried to write PNG data past end of buffer");
+    }
+    memcpy(&e->buf[e->write_pos], data, length);
+    e->write_pos += length;
+}
+
+void user_flush_data(png_structp png_ptr) {
+    // Do nothing, since this isn't actually a file
+}
+
+apng_decoder apng_decoder_create(const opencv_mat buf) {
+    uint32_t width, height;
+    apng_decoder d = new struct apng_decoder_struct();
+    memset(d, 0, sizeof(struct apng_decoder_struct));
+    d->mat = static_cast<const cv::Mat*>(buf);
+
+    if (png_sig_cmp(d->mat->data, 0, 8) != 0)
+        goto error;
+
+    d->png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+    if (!d->png_ptr)
+        goto error;
+
+    d->info_ptr = png_create_info_struct(d->png_ptr);
+    if (!d->info_ptr)
+        goto error;
+
+    if (setjmp(png_jmpbuf(d->png_ptr))) {
+        goto error;
+    }
+
+    png_set_read_fn(d->png_ptr, d, user_read_data);
+
+    png_read_info(d->png_ptr, d->info_ptr);
+
+    // Apply transformations to image data
+    png_set_expand(d->png_ptr);
+    png_set_strip_16(d->png_ptr);
+    png_set_gray_to_rgb(d->png_ptr);
+    png_set_add_alpha(d->png_ptr, 0xFF, PNG_FILLER_AFTER);
+    png_set_bgr(d->png_ptr);
+
+    png_read_update_info(d->png_ptr, d->info_ptr);
+
+    if(!png_get_valid(d->png_ptr, d->info_ptr, PNG_INFO_acTL))
+        goto error;
+
+    width = png_get_image_width(d->png_ptr, d->info_ptr);
+    height = png_get_image_height(d->png_ptr, d->info_ptr);
+
+    d->prev_frame = (uint8_t *) calloc(4, width * height);
```

Comment:
One thing I'm wondering on this - in `ops.go` we already toggle between two pre-allocated frame buffers. What would you think about just taking a `previous frame` argument for the decoding below? Then the caller would be responsible for passing previous frame, and we could avoid the allocations?